### PR TITLE
Use an interface for tags.

### LINKED ISSUE #105: pubsub: use an interface with Get() and Len() for tags
rather than a map

### PR DESCRIPTION
**Pull request by tomc974**
Fixes #105

The issue was a little vague, so I'm not sure the fix is quite  what was expected.
Also  I could not come with a better name than `ITagSet` which is not quite the Go way.

### CODE REVIEW

**Comment by melekes** [comment]
@tomc974 thank you for working on this!

File: pubsub/pubsub.go [hunk 1]
```diff
@@ -38,18 +38,25 @@ var (
 	ErrAlreadySubscribed = errors.New("already subscribed")
 )
 
+type ITagSet interface {
```
Comment (melekes) [inline pubsub/pubsub.go]: how about

```
// TagMap ....
type TagMap interface {
        // Get returns the value for a key, or nil if no value is present. The ok result indicates whether value was found in the tags.
        Get(key string) (value interface{}, ok bool)
        // Len returns the number of tags
	Len() int
}

type tagMap map[string]interface{}

....
```

File: pubsub/pubsub.go [hunk 2]
```diff
@@ -38,18 +38,25 @@ var (
...
+	Get(key string) (result interface{}, found bool)
+	Len() int
+}
+
+type TagSet map[string]interface{}
```
Comment (melekes) [inline pubsub/pubsub.go]: do we need to export this?? see my comment above

File: pubsub/pubsub.go (hunk 1 above)
Comment (tomc974) [reply pubsub/pubsub.go]: Will do that!

File: pubsub/pubsub.go (hunk 2 above)
Comment (tomc974) [reply pubsub/pubsub.go]: Don't know what it is used for but if you think it should be private so be it.

File: pubsub/pubsub.go [hunk 3]
```diff
@@ -38,18 +38,28 @@ var (
 	ErrAlreadySubscribed = errors.New("already subscribed")
 )
 
+type TagMap interface {
```
Comment (melekes) [inline pubsub/pubsub.go]: could you please add a comment

Comment (tomc974) [reply pubsub/pubsub.go]: done!

**Comment by melekes** [comment]
Wanna upgrade Tendermint to support this?

**Comment by tomc974** [comment]
Yeah sure, I'll do that.

**Pull request merged**